Add migrate to schema v2 skill (#55)
open-telemetry/opentelemetry-weaver-packages: skills/migrate-to-schema-v2.md

---
name: migrate-to-definitions-schema-v2

> description: 'Use when writing semantic conventions v2 definition schema (file_format: definition/2*) or migrating v1 schema files to v2 for entities, spans, metrics, events, attribute groups, or signal refinements in any semantic-conventions registry. Enforces minimal attribute groups, internal-by-default visibility, flat structure, and refinement of generic signals over re-declaration.

Reviewer: The description in the semantic-conventions copy is narrower (migration-focused) and doesn't advertise `definition/2*`:

```suggestion
description: 'Use when migrating v1 schema files to v2 (file_format: definition/2)
for entities, spans, metrics, events, attribute groups, or signal refinements in
any semantic-conventions registry. Enforces minimal attribute groups,
internal-by-default visibility, flat structure, and refinement of generic signals
over re-declaration.
```

---

# Migrate semantic conventions from current definition schema to v2

Use this skill when migrating from existing semantic convention (v1, `groups`)
defining attributes, spans, metrics, events, entities, attribute groups, or signal refinement;

> to v2 (`file_format: definition/2` or `file_format: definition/2.*`).

Reviewer: `definition/2.*` is not accepted by weaver - it rejects `definition/2.0.0` and other suffixed forms:

```suggestion
to v2 (`file_format: definition/2` — that exact string; weaver rejects `definition/2.0.0` and other forms).
```

Can also be used for new v2 semconv definitions.

## Goal

Produce a v2 definition that is minimal, flat, and resolvable by weaver

> without errors.

Reviewer: ```suggestion
without errors from existing valid v1 semantic conventions.
```

This skill is not for deciding *whether* a convention should be added or
arguing about attribute names, requirement levels, or stability — those
are upstream design questions. It is also not a generator: bring the
names, descriptions, and examples; the skill shapes the structure around
them.

## Schema Reference

This skill encodes opinions, not the v2 grammar. For top-level sections,
required keys per signal, valid `kind` / `instrument` / `stability` /
`requirement_level` values, and override rules — read the doc directly:

- [Lifecycle and overview](https://raw.githubusercontent.com/open-telemetry/weaver/main/schemas/semconv-schemas.md)
- [Authoring syntax reference](https://raw.githubusercontent.com/open-telemetry/weaver/main/schemas/semconv-syntax.v2.md)
- [Reference of an old syntax](https://raw.githubusercontent.com/open-telemetry/weaver/main/schemas/semconv-syntax.md)

> If anything here contradicts these docs, docs win.

Reviewer: The JSON Schema is the authoritative reference and is missing here; the prose doc alone gets property-level rules wrong. Also add the span `name` requirement note:

```suggestion
This skill encodes opinions, not the v2 grammar. Before writing any YAML,
read the authoring references **in full** — do not infer the grammar from
summaries or from surrounding files (summaries get the property-level rules
wrong, e.g. which `ref` properties are valid where):

- [Authoring JSON Schema](https://raw.githubusercontent.com/open-telemetry/weaver/main/schemas/semconv.schema.v2.json)
  — authoritative and machine-precise: top-level sections, required keys per
  signal, valid `kind` / `instrument` / `stability` / `requirement_level`
  values, and exactly which properties each attribute `ref` may carry. Note
  `sampling_relevant` lives only on `SpanAttributeRef` — it is not valid on
  attribute-group refs. Read this first.
- [Authoring syntax reference (prose)](https://raw.githubusercontent.com/open-telemetry/weaver/main/schemas/semconv-syntax.v2.md)
  — narrative for the same grammar, plus lifecycle/override behavior.
- [Lifecycle and overview](https://raw.githubusercontent.com/open-telemetry/weaver/main/schemas/semconv-schemas.md)
- [Reference of an old syntax](https://raw.githubusercontent.com/open-telemetry/weaver/main/schemas/semconv-syntax.md)

If anything here contradicts these, they win — the JSON Schema first.

v2 spans require a `name` block. If the source gives no span-name guidance,
that's an authoring decision — surface it, don't invent normative text.
```

## Weaver Commands

Paths (`./model`, `./templates`, etc.) are examples — adjust to the actual registry layout.

| Goal | Command |
|------|---------|
| Validate and package | `weaver registry package -r ./model --v2` |
| Check (no packaging) | `weaver registry check -r ./model --v2` |
| Check policies | `weaver registry check -r ./model --v2 -p https://github.com/open-telemetry/opentelemetry-weaver-packages.git[policies/check/naming_conventions]` |
| Generate registry docs | `weaver registry generate -r ./model --v2 -t ./templates/registry markdown ./docs/registry` |
| Refresh doc snippets | `weaver registry update-markdown -r ./model --v2 -t ./templates --target markdown docs` |

> `weaver registry resolve` is deprecated — always use `registry package` for validation.

Multiple `-p` flags can be combined in a single `check` invocation. When this skill says **package**, **check**, **generate-registry**, or **generate-docs**, it means the matching command above.

> > **v2 compatibility:** Policies and templates must support v2. The policies in this repo ([opentelemetry-weaver-packages](https://github.com/open-telemetry/opentelemetry-weaver-packages)) do. For doc generation, confirm the project's templates are v2-aware before running **generate-registry** or **generate-docs**. If the templates are v1-only: skip doc regeneration steps in Procedure and Output Format, and note to the user that comparing **package** output before and after the change is still valid and sufficient for validation. Suggest adopting v2 templates from [semantic-conventions-genai/templates](https://github.com/open-telemetry/semantic-conventions-genai/tree/main/templates) as a starting point.

Reviewer: Missing the `--future` gotcha (doc generation fails on `definition/2` when `--future` is passed), and the v2-compatibility note can be much shorter:

```suggestion
> `weaver registry resolve` is deprecated — always use `registry package` for validation.

> **Do not pass `--future` for v2 doc generation.** `update-markdown` and `generate` reject a `definition/2` file with *"File format definition/2 is not yet stable"* when `--future` is set — omit it (and drop it from the repo's Makefile doc-gen targets if present). `package` is unaffected.

Multiple `-p` flags can be combined in a single `check` invocation. When this skill says **package**, **check**, **generate-registry**, or **generate-docs**, it means the matching command above.

> **v2 compatibility:** Policies and templates must be v2-aware (this repo's [policies](https://github.com/open-telemetry/opentelemetry-weaver-packages) are). If doc generation can't run, a `package`/`check` before-vs-after diff is still sufficient validation.
```

## Authoring Rules

### 1. Minimal attribute groups

Only create an `attribute_groups` entry when the *same exact set* of
attributes is reused by two or more signals. A group used by exactly
one signal is dead weight — inline the attributes on the signal

> directly.

Reviewer: The rule needs the second gate too (a group must add ≥ 2 attributes of its own), and the count must include `*_refinements` sites:

```suggestion
An `attribute_groups` entry earns its place only if it passes **both** of
these tests. Fail either one → inline it and delete the group.

- **Used ≥ 2 times.** Count *every* referencing site across the whole
  registry, including all four `*_refinements` sections — not just `metrics` /
  `spans` / `events` / `entities`. A group
  referenced by a single signal *or* a single refinement is used once.
- **Adds ≥ 2 attributes of its own.** Count only the `ref:` entries the
  group declares beyond any `ref_group:` it wraps. A group that is
  `ref_group: base` + one `ref:` adds one attribute; the indirection and
  extra nesting level cost more than the one line it saves at each site,
  even when many signals share it.

Inlining means repeating `ref_group: <base>` plus the group's own `ref:`
entries directly on each signal/refinement. This produces the identical
resolved attribute **set**. Always verify with a diff (Procedure step 8).

When in doubt, inline: reviewers ask you to delete a near-empty group far
more often than they ask you to introduce one.
```

### 2. Visibility: structural vs. documented groups

The right visibility depends on whether the group carries documentation fields:

- **Structural groups** (no `brief` or `stability`, exist only to bundle `ref:` entries
  for use by signals) → `visibility: internal`.
- **Documented groups** (carry `brief` and/or `stability` to appear in generated docs,
  e.g. a reusable "these attributes may be used for X" group) → `visibility: public`.
  The schema enforces this: `check-policies` will reject `internal` when `brief` or
  `stability` are present.

Default to `internal` when writing a new structural group. Promote to `public` only
when the group needs a `brief`/`stability` of its own — that is a deliberate authoring
decision, not a default.

**Side-effect when migrating:** Promoting a group to `public` with a `brief` causes
the template to emit that description inside every `<!-- semconv <group-id> -->` block
in hand-authored docs. If a human already wrote that same text above the block, the

> description will be duplicated after regeneration. See Procedure step 5a.

Reviewer: This contradicts the semantic-conventions version, which is internal-only and leaves `public` as an explicit human decision:

```suggestion
### 2. Internal by default

Every authoring `attribute_groups` entry MUST have `visibility: internal`.
Decision to make group public must be a human decision. Don't create public groups.
```

### 3. Flat structure

The [upstream syntax doc](https://github.com/open-telemetry/weaver/blob/main/schemas/semconv-syntax.v2.md#attribute-group-reference)

> says it is *NOT RECOMMENDED* to use `ref_group` on another attribute

Reviewer 2: Should we modify upstream docs?

group. This is not always feasible and conflicts with minimalism and avoiding repetitions.

Do the middle-ground, keep the chain shallow:

- **One level** (`base group → signal`) is the default.
- **Two levels** (`base → composite → signal`) only when multiple signals
  share the same composite shape and inlining would duplicate briefs, notes, or other substantial blocks.
  Justify in review; do not add speculatively.
- **Three+ levels** — almost never. Treat as a refactor signal.

Within an `attributes:` list, list every `ref_group` entry before any
`ref` entry — the group's own attributes appear after the inherited

> ones, not interleaved.

Reviewer: Not present in the semantic-conventions version - either drop it or confirm it's a deliberate addition here (and that resolved order actually behaves this way).

```yaml
attribute_groups:
  - id: attributes.<domain>.common         # level 1
    visibility: internal
    attributes:
      - ref: <attr.a>

  - id: attributes.<domain>.client         # level 2
    visibility: internal
    attributes:
      - ref_group: attributes.<domain>.common
      - ref: <attr.b>
```

### 4. Examples must be lists

In v2, `examples` values must always be YAML sequences, even for a single value.
Scalar literals from v1 must be wrapped:

```yaml
# v1
examples: 42
examples: main

# v2
examples: [42]
examples: ['main']

> ```

Reviewer: "Examples must be lists" doesn't exist in the semantic-conventions version. If it stays, the rule numbering below is broken - it jumps 4 → 6 → 7 → 8, so the remaining rules should be 5/6/7 and the cross-references updated.

> ### 6. Refinement over redefinition

Reviewer: ```suggestion
### 5. Refinement over redefinition
```

If the new signal is an implementation-specific variant of a generic
signal that already exists in this registry, declare it in the matching
`*_refinements` section and add only the delta.

Refinements inherit the parent's attributes — do **not** re-list them.
List an inherited attribute only to override its presentation (note,
brief, requirement_level, examples, sampling_relevant — see the syntax
doc for which keys are overridable per signal type).

Tempted to declare the variant in `spans:` and `ref_group:` the parent's
attribute group(s)? Stop and use a refinement. Same applies to implicit

> signal groups (`ref_group: span.<parent>`) — see Rule 6.

Reviewer: `see Rule 6` points at itself. It should point at the implicit-signal-groups rule. Also missing the "refinement is a leaf" exception from the semantic-conventions version:

```suggestion
Tempted to declare the variant in `spans:` and `ref_group:` the parent's
attribute group(s)? Stop and use a refinement. Same applies to implicit
signal groups (`ref_group: span.<parent>`) — see Rule 7.

Exception: if the variant will itself be refined downstream, keep it as a
signal and share attributes via a group — see step 0 of the decision tree.
```

> ### 7. Override on the signal, not on the attribute

Reviewer: ```suggestion
### 6. Override on the signal, not on the attribute
```

through. Use `note: ""` only when you genuinely want to suppress an
inherited note.

### 8. Implicit signal groups

Weaver exposes each signal's full attribute set — every override and
every inlined ref — as an implicit group named `span.<type>`,
`metric.<name>`, or `event.<name>`. This is undocumented in the upstream
syntax doc; treat as undocumented behavior and verify with **[package](#weaver-commands)**
whenever you rely on it.

Legitimate use: **cross-signal mirroring**, e.g.
`gen_ai.client.inference.operation.details` (an event) uses
`ref_group: span.gen_ai.inference.client` so it picks up every override
the span declaration carries. Re-stating each `ref_group` and override
would drift the moment the span changes.

Do **not** use this inside a new signal under
`spans:` / `metrics:` / `events:` to clone a generic signal — that's
Rule 4's anti-pattern in different syntax. Use `*_refinements` instead.
Symptoms that you are about to drift:

- A `spans:` entry whose `attributes:` is `ref_group: span.<other_type>`
  followed by a small delta.
- A new internal `attributes.<provider>.<thing>` group whose only purpose

>   is to hold that delta.

Reviewer: This section is much longer than the semantic-conventions version and hardcodes a gen_ai-specific example, which doesn't generalize for a shared skill:

```suggestion
### 7. Implicit signal groups

Weaver exposes each signal's resolved attribute set as an implicit group
`span.<type>` / `metric.<name>` / `event.<name>` (undocumented — verify with
**[package](#weaver-commands)**). Legitimate use is cross-signal mirroring: an
event reusing a span's set via `ref_group: span.<type>`. Do **not** use it to
clone a generic signal into a new `spans:` / `metrics:` / `events:` entry —
that's Rule 5's anti-pattern; use `*_refinements`.
```

## Refinement Decision Tree

For every new signal, walk this before writing yaml:

1. **Implementation-specific variant of an existing signal?**
   → declare it in `*_refinements` with `ref:` to the parent. Delta only.
   Per-attribute overrides for any inherited attribute whose framing
   changes. Done.
2. **Shares its attribute set with a signal already in the registry?**
   → reuse the existing group via `ref_group`. Delta inline on the signal.
3. **Shares its attribute set with another signal in this same registry?**
   → one shared internal group. One level deep. No nesting.

> 4. **None of the above?** → inline the attributes on the signal. No group.

Reviewer: Missing step 0 (a refinement is a leaf - if downstream registries will refine it, keep it a signal). Steps 2 and 3 also read as duplicates of each other:

```suggestion
For each new signal, before writing YAML:

0. Will other registries or system-specific conventions refine this signal? →
   declare it as a signal (`spans:` / `metrics:` / `events:` / `entities:`) and
   share attributes via a group. A refinement is a leaf — don't build on top of one.
1. Variant of an existing signal? → `*_refinements`, `ref:` the parent, delta only (Rule 5).
2. Same attribute set as an existing group? → `ref_group` it, delta inline.
3. Shares an attribute set with another new signal here? → one shared internal group, one level (Rule 1).
4. Otherwise → inline on the signal, no group.
```

## Procedure

1. Identify what the change touches: spans, metrics, events, entities,
   attribute groups, refinements, or several.
2. **Capture a pre-change baseline.** Run **[package](#weaver-commands)** so the resolved-schema
   artifact reflects HEAD; you'll diff against this after editing.
   Skip if the user mentioned a baseline or the working tree is in sync.
3. If syntax for any touched section isn't obvious from surrounding

>    files, fetch the v2 syntax reference (see Schema Reference).

Reviewer: "if syntax isn't obvious" is weaker than the semantic-conventions version, which requires reading the references up front:

```suggestion
3. Read the authoring references in full (see Schema Reference) before
   writing YAML — the JSON Schema is authoritative for what each section and
   attribute `ref` may carry. Don't infer the grammar from summaries or
   surrounding files.
```

4. Walk the Refinement Decision Tree for each new signal.
5. Write the yaml; satisfy the rules.
   - **5a. Documented-group duplicate check.** For every group you are
     migrating to `visibility: public` with a `brief`, grep `docs/` for
     `<!-- semconv <group-id> -->` and read the lines immediately before
     each hit. If there is hand-authored prose that paraphrases or copies
     the `brief`, remove it — the template will now emit it inside the
     autogenerated block, and leaving it outside creates a duplicate after

>      regeneration.

Reviewer: Step 5a exists only because of the public/documented-group rule above. Drop it along with Rule 2's public-group branch:

```suggestion
```

6. Run validation (next section). Fix every error.
7. Re-read the diff for: groups used exactly once, `ref_group` chains
   deeper than two levels, `visibility: public` on a structural group that
   has no `brief`/`stability` (should be internal), `visibility: internal`
   on a documented group that carries `brief`/`stability` (policy will
   reject it), re-listed parent attributes on refinements,
   `ref_group: span.<...>` inside a new signal declaration. These are the

>    common drifts.

Reviewer: The Rule 1 gate should be checked explicitly here (grep + count both references and own refs), and the public/internal branches go away with Rule 2:

```suggestion
7. Re-read the diff for the common drifts. For **every**
   `attribute_groups` entry, run the Rule 1 gate explicitly: grep the
   registry for its id and count references (include `*_refinements`), and
   count the `ref:` entries it adds beyond its `ref_group:`. Fewer than 2
   of either → inline and delete. Also check: `ref_group` chains deeper
   than two levels, `visibility: public` that should be internal,
   re-listed parent attributes on refinements, `ref_group: span.<...>`
   inside a new signal declaration.
```

8. Regenerate every committed artifact (see Output Format) and compare
   against the step-2 baseline.

## Validation

The non-negotiable gate: the resolved registry builds cleanly with zero
errors. Run **[package](#weaver-commands)**.

If a **[check](#weaver-commands)** command is configured with policies, run it — it catches
id collisions and naming-convention violations that `package` alone

> does not.

Reviewer: Missing the backward-compatibility subsection - important because turning a released signal into a refinement trips the policy, and the fix is a spec decision, not something the agent should work around:

```suggestion
does not.

### Backward compatibility

A migration is a refactor: the backwards-compatibility policy should pass
unchanged. Converting a released signal into a refinement removes its own
type/name from the resolved registry and will trip it.

If it fails, do not work around it on your own — deprecating the old signal,
renaming ids, or exempting it in policy are all spec decisions. Report what
broke and let a human decide whether an exemption is appropriate.
```

## Output Format

Do not write a prose summary of what changed. Link the artifacts. Run in order — docs both consume
the resolved model that `package` produces:

1. **[package](#weaver-commands)** — builds the resolved schema and runs validation.
   Link the resolved schema file.
2. **Docs regeneration** — **[generate-registry](#weaver-commands)** rebuilds the
   per-namespace pages under `docs/registry/`; **[generate-docs](#weaver-commands)**
   refreshes the `<!-- semconv ... -->` snippet tables in hand-authored
   docs. Other registries may combine these under one command.

After running, link every file the regeneration changed:

> regenerated `docs/registry/` pages, and any snippet refreshes elsewhere

Reviewer: The semantic-conventions version is simpler here - `package` already ran in Validation, so re-listing it is redundant, and it asks for a short recap rather than "no prose summary":

```suggestion
After validation passes, regenerate the committed docs:

1. **[generate-registry](#weaver-commands)** — rebuilds the per-namespace pages under `docs/registry/`.
2. **[generate-docs](#weaver-commands)** — refreshes the `<!-- semconv ... -->` snippet tables in hand-authored docs.

Some registries combine these or wrap them in a Makefile target (e.g. `table-generation`/`table-check`).
Then recap concisely: what changed and which files were regenerated — don't re-narrate every edit.
```
